[WIP] SyliusReviewBundle

Zales0123 opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Fixed tickets |  |
| License | MIT |
| Doc PR | in progress |

Initial implementation of SyliusReviewBundle, which allows to easily manage review system. It's based on https://github.com/Sylius/Sylius/pull/431. "Reviewable" and "Reviewer" classes can be configured under `sylius_review` tag:

``` yml
sylius_review:
    classes:
        product:
            subject: Sylius\Component\Core\Model\Product
            review:
                model: Sylius\Component\Product\Model\ProductReview
                form:
                    default: Sylius\Bundle\ProductBundle\Form\Type\ProductReviewType
                    admin: Sylius\Bundle\CoreBundle\Form\Type\ProductReviewAdminType
            reviewer:
                model: Sylius\Component\Core\Model\Customer
```

Reviews can be written by customers and guests on the product page. Newly created review has state `new` and can be `confirmed` or `rejected` by administrator. After confirming, it's product has average rating calculated. There are 3 possible behaviors while review creation:
- if customer is logged in, it's automatically assigned as review author
- if customer is not logged in, email must be provided and new customer is created based on it
- if provided email is already used, "please, log in" message is displayed

Reviews can be also created in administration panel, with choosing product and author.

There are still some things that have to be changed to make it working perfectly:
- [ ] change ProductReviewType to use CustomerGuestType or some similar UserBundle form type: for now there is custom transformer and listener, cause existing CustomerGuestType doesn't exactly fit ReviewBundle needs - they would probably removed after fixing this point
- [ ] consider how review author should be displayed on reviews list:
  - john.doe@example.com
  - John Doe
  - John D.
- [ ] think about CustomerDeleteListener: currently every review of deleted customer is deleted as well, but I think there should have author field changed to "deleted user" or sth like this - I would appreciate any ideas about implementation
- [ ] work on translations, cause I'm sure not everything is translated for now

Some ideas for future improvements:
- [ ] send notification email to review author after accepting/rejecting his review
- [ ] add possibility to send reason, why review was rejected
- [ ] add proper info label to review if authored by user, that have already bought reviewed subject

I will be thankful for some review, cause I'm pretty sure there are some part when my code can be improved ;)

Richtermeister committed:
[Review] SyliusReviewBundle initialization

Zales0123 committed:
[Review] ReviewBundle refactoring

Split bundle to bundle and component.
Removed unnecessary classes, due to logic changes.
General code refactoring.

Zales0123 committed:
[Review] Basic review logic finished, with added validation

Zales0123 committed:
[Review] Behat scenarios and average rating feature implemented

pamil commented on features/frontend/products.feature:
```diff
@@ -9,17 +9,20 @@ Feature: Products
-            | Sylius Tee       | 12.99 | PHP T-Shirts |
-            | Symfony T-Shirt  | 15.00 | PHP T-Shirts |
-            | Doctrine T-Shirt | 15.00 | PHP T-Shirts |
+            | name             | price | taxons       | average_rating |
```
There should be space instead of `_`, it's business scenario :)

pamil commented on features/reviews/reviews.feature:
```diff
@@ -0,0 +1,86 @@
+
+    Background:
+        Given there are following users:
+            | email          | enabled  | created at          |
```
I may be wrong, but isn't user enabled by default?

pamil replied:
Anyway, created at shouldn't be here.

pamil commented on features/reviews/reviews.feature:
```diff
@@ -0,0 +1,86 @@
+            | Super T-Shirt    | 19.99 | T-Shirts     | 0              |
+            | Black T-Shirt    | 18.99 | T-Shirts     | 0              |
+            | Sylius Tee       | 12.99 | PHP T-Shirts | 0              |
+            | Symfony T-Shirt  | 15.00 | PHP T-Shirts | 5              |
```
Why it's avergae rating is 5 if there's only one review with lowest rating possible? (Same comes with other average ratings in this table)

pamil commented on features/reviews/reviews.feature:
```diff
@@ -0,0 +1,86 @@
+            | Comment | Lorem ipsum dolor |
+          And I select the "5" radio button
+          And I press "Create"
+         Then I should be on the page of review "Report2"
```
What's `Report2`?

pamil commented on features/reviews/reviews.feature:
```diff
@@ -0,0 +1,86 @@
+
+    Scenario: Removing review from the list
+        Given I am on the review index page
+         When I press "Delete" near "Awesome"
```
We use modals for every delete action in Sylius.

pamil commented on features/reviews/reviews.feature:
```diff
@@ -0,0 +1,86 @@
+
+    Scenario: Removing review from details page
+        Given I am on the review "Awesome" page
+         When I press "delete"
```
We use modals for every delete action in Sylius.

pamil commented on src/Sylius/Bundle/ReviewBundle/DependencyInjection/Configuration.php:
```diff
@@ -0,0 +1,93 @@
+                            ->children()
+                                ->scalarNode('model')->defaultValue('Sylius\\Component\\Review\\Model\\Review')->end()
+                                ->scalarNode('controller')->defaultValue('Sylius\\Bundle\\ReviewBundle\\Controller\\ReviewController')->end()
+                                ->scalarNode('repository')->cannotBeEmpty()->end()
```
Why there's no default one defined?

pamil commented on src/Sylius/Bundle/ReviewBundle/EventListener/ReviewCreateListener.php:
```diff
@@ -0,0 +1,87 @@
+    public function __construct(
+        AverageRatingCalculatorInterface $averageRatingCalculator,
+        CustomerContextInterface $customerContext,
+        ObjectManager $productManager
```
You don't know its product manager yet, it's `$reviewSubjectManager`.

pamil commented on src/Sylius/Bundle/ReviewBundle/Form/Type/ReviewAdminType.php:
```diff
@@ -0,0 +1,62 @@
+        $builder
+            ->remove('author')
+            ->add('author', 'entity', array(
+                'class' => 'Sylius\Component\Core\Model\Customer',
```
You're still in ReviewBundle, not CoreBundle.

pamil commented on src/Sylius/Bundle/ReviewBundle/Form/Type/ReviewAdminType.php:
```diff
@@ -0,0 +1,62 @@
+                'label' => 'sylius.form.review.status.label'
+            ))
+            ->add('product', 'entity', array(
+                'class'    => 'Sylius\Component\Core\Model\Product',
```
Same here.

pamil commented on src/Sylius/Bundle/ReviewBundle/Form/Type/ReviewType.php:
```diff
@@ -0,0 +1,84 @@
+    {
+        $builder
+            ->add('rating', 'choice', array(
+                'choices' => $this->createRatingList($options['rating_steps']),
```
`range(1, 5)`

pamil commented on src/Sylius/Bundle/ReviewBundle/README.md:
```diff
@@ -0,0 +1,72 @@
+Authors
+-------
+
+The bundle was originally created by [Paweł Jędrzejewski](http://pjedrzejewski.com).
```
Are you sure? :)

pamil commented on src/Sylius/Bundle/ReviewBundle/Resources/config/doctrine/model/Review.orm.xml:
```diff
@@ -0,0 +1,42 @@
+        </id>
+        <field name="title" type="string" column="title"/>
+        <field name="rating" type="integer" column="rating"/>
+        <field name="comment" type="text" column="comment"/>
```
I think we should allow reviews without comment, as many people will want to rate something, not to write an opinion.

pamil replied:
Anyway, it would be awesome to have feature, that checks whether review is verified (eg. was sent by customer, that had bought that product earlier).

pamil commented on src/Sylius/Bundle/ReviewBundle/Resources/config/doctrine/model/Review.orm.xml:
```diff
@@ -0,0 +1,42 @@
+            </cascade>
+            <join-column name="author_id" referenced-column-name="id" />
+        </many-to-one>
+        <many-to-one field="product" target-entity="Sylius\Component\Product\Model\ProductInterface" inversed-by="reviews">
```
It's still ReviewBundle, we don't need coupling with Product here.

pamil commented on src/Sylius/Bundle/ReviewBundle/Validator/Constraints/UniqueCustomerEmailValidator.php:
```diff
@@ -0,0 +1,51 @@
+
+namespace Sylius\Bundle\ReviewBundle\Validator\Constraints;
+
+use Doctrine\ORM\EntityRepository;
```
Let's use `ObjectRepository`, it is less specific and still fits our use-case.

pamil commented on src/Sylius/Bundle/ReviewBundle/composer.json:
```diff
@@ -0,0 +1,54 @@
+        "symfony/framework-bundle":        "~2.3",
+        "stof/doctrine-extensions-bundle": "~1.1",
+        "sylius/resource-bundle":          "0.15.*@dev",
+        "sylius/money-bundle":             "0.15.*@dev",
```
Why?

pamil commented on src/Sylius/Bundle/ReviewBundle/composer.json:
```diff
@@ -0,0 +1,54 @@
+        "stof/doctrine-extensions-bundle": "~1.1",
+        "sylius/resource-bundle":          "0.15.*@dev",
+        "sylius/money-bundle":             "0.15.*@dev",
+        "sylius/report":                   "0.15.*@dev"
```
Why?

pamil commented on src/Sylius/Bundle/ReviewBundle/spec/EventListener/ReviewCreateListenerSpec.php:
```diff
@@ -0,0 +1,102 @@
+
+        $review->setAuthor($customer)->shouldBeCalled();
+
+        $this->controlReviewAuthor($event);
```
I think `control` isn't the best verb here. What do you think about `ensureReviewHasAuthor()`?

pamil commented on src/Sylius/Bundle/ReviewBundle/spec/EventListener/ReviewCreateListenerSpec.php:
```diff
@@ -0,0 +1,102 @@
+
+        $customerContext->getCustomer()->shouldNotBeCalled();
+
+        $this->controlReviewAuthor($event)->shouldReturn(null);
```
To be specific, `controlReviewAuthor` returns nothing (`void`). It is not `null`, but due to PHP5 weirdness the spec passes :) IMHO there shouldn't be `shouldReturn` at all.

pamil commented on src/Sylius/Bundle/ReviewBundle/Form/Transformer/ReviewAuthorTransformer.php:
```diff
@@ -0,0 +1,54 @@
+            return;
+        }
+
+        $customer = new Customer();
```
Why didn't you use factory?

pamil commented on src/Sylius/Bundle/ReviewBundle/spec/Form/Transformer/ReviewAuthorTransformerSpec.php:
```diff
@@ -0,0 +1,76 @@
+                if (!$subject instanceof CustomerInterface || !$key instanceof CustomerInterface) {
+                    return false;
+                }
+                return (
```
Not necessary brackets. Also, there should be blank line before `return` statement.

pamil commented on src/Sylius/Bundle/ReviewBundle/spec/Validator/Constraints/UniqueCustomerEmailSpec.php:
```diff
@@ -0,0 +1,46 @@
+
+    function it_has_targets()
+    {
+        $this->getTargets()->shouldReturn('class');
```
There's no reason to hardcode targets, let's use `Constraint` constants.

pamil commented on src/Sylius/Bundle/WebBundle/Controller/Frontend/Review/ProductReviewController.php:
```diff
@@ -0,0 +1,163 @@
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+
+/**
+ * ProductReview controller.
```
![you_dont_say](https://cloud.githubusercontent.com/assets/1897953/9438901/77991e88-4a62-11e5-96c1-783ac5eca6d2.png)

pamil commented on src/Sylius/Bundle/WebBundle/Controller/Frontend/Review/ProductReviewController.php:
```diff
@@ -0,0 +1,163 @@
+     *
+     * @return array
+     */
+    private function getDeepErrors(FormInterface $form)
```
If it's scheduled for `0.16` Sylius version, you can use `$form->getErrors(true)` instaed of this method.

pamil commented on src/Sylius/Bundle/WebBundle/Resources/views/Backend/Review/index.html.twig:
```diff
@@ -0,0 +1,27 @@
+{% block content %}
+<div class="page-header">
+    <div class="actions-menu">
+        {{ buttons.create(path('sylius_backend_review_create'), 'Create Review') }}
```
Translations! :)

pamil commented on src/Sylius/Bundle/WebBundle/Resources/views/Backend/Review/macros.html.twig:
```diff
@@ -0,0 +1,51 @@
+        </tbody>
+    </table>
+{% else %}
+{{ alerts.info('There are no reviews. ') }}
```
Translations :)

pamil commented on src/Sylius/Bundle/WebBundle/Resources/views/Backend/Review/macros.html.twig:
```diff
@@ -0,0 +1,51 @@
+    <table class="table table-hover">
+        <thead>
+            <tr>
+                <th>ID</th>
```
Translations

pamil commented on src/Sylius/Bundle/WebBundle/Resources/views/Frontend/Product/show.html.twig:
```diff
@@ -121,4 +122,56 @@
+
+{% block javascripts %}
+    {{ parent() }}
+    <script>
```
Inline JavaScript is the root of all evil (also missing `type` attribute :D).

pamil commented on src/Sylius/Component/Core/Model/Product.php:
```diff
@@ -16,42 +16,33 @@
 use Sylius\Component\Taxonomy\Model\TaxonInterface as BaseTaxonInterface;
 
 /**
- * Sylius core product entity.
```
Good job! :+1:

pamil commented on src/Sylius/Bundle/WebBundle/Resources/views/Frontend/Review/_list.html.twig:
```diff
@@ -0,0 +1,16 @@
+            </h4>
+            Rating: {{ review.rating }}
+            <br>
+            Author: {{ review.author.email }}
```
[RFC] I think many people would complain about their email being publicly available. We can use "John D." for "John Doe", it's quite anonymous. Any other ideas? @pjedrzejewski

pamil commented on src/Sylius/Component/Review/Calculator/AverageRatingCalculator.php:
```diff
@@ -0,0 +1,38 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function calculate(Reviewable $reviewable)
```
We will need more efficient way to calculate average rating. Calculating it by sum of ratings and amount of them should do the job.

pamil commented on src/Sylius/Component/Review/Model/Review.php:
```diff
@@ -0,0 +1,212 @@
+    /**
+     * @param CustomerInterface $customer
+     */
+    public function setAuthor(CustomerInterface $customer = null)
```
Let's use `$author` here, IMHO it's more appropriate and clear.

pamil commented on src/Sylius/Component/Review/Model/Review.php:
```diff
@@ -0,0 +1,212 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function getProduct()
```
There's no product yet, you're still in `Review` component :)

pamil commented on src/Sylius/Component/Review/Model/Reviewable.php:
```diff
@@ -0,0 +1,50 @@
+    /**
+     * @param Collection $reviews
+     */
+    public function setReviews(Collection $reviews);
```
Do we really need it here?

pamil commented on src/Sylius/Component/Review/Model/Reviewable.php:
```diff
@@ -0,0 +1,50 @@
+    /**
+     * @param float $averageRating
+     */
+    public function setAverageRating($averageRating);
```
I'm not sure about storing average rating in `Reviewable` object, it feels like rating should not be property of `Reviewable`.

pamil commented on src/Sylius/Component/Review/Model/Reviewable.php:
```diff
@@ -0,0 +1,50 @@
+/**
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+interface Reviewable
```
Symfony2 codings standards (or PSR-2 too, I'm not sure) says, that there must be `Interface` suffix after interface name.

pamil commented on src/Sylius/Component/Review/spec/Model/ReviewSpec.php:
```diff
@@ -0,0 +1,80 @@
+
+    function it_has_rating()
+    {
+        $this->setRating(5);
```
It is 0 by default, it's worth the another assertion :)

pamil commented on src/Sylius/Component/Review/spec/Model/ReviewSpec.php:
```diff
@@ -0,0 +1,80 @@
+
+    function it_has_status()
+    {
+        $this->setStatus('new');
```
Isn't it the value of a constant?

Zales0123 committed:
[Review] State machine for review implemented

Calculate average rating on review confirmation, not creation.

Zales0123 committed:
[Review] General fixes and improvements

Zales0123 committed:
[Review] General fixes and improvements

Zales0123 committed:
[Review] Make reviews more generic

Split configurations into Review-, Product- and CoreBundle parts

michalmarcinkowski committed:
Update requirements and minor fixes

Zales0123 committed:
[Review] Split ProductReviewAdminType to Product- and CoreBundle part

Zales0123 committed:
[Review] Implemented ReviewDeleteListener

Also some general and CS fixes.

Zales0123 committed:
[Review] Change ReviewAuthorInterface to ReviewerInterface

Zales0123 committed:
[Review] Added default Reviewer class

HerbMat committed:
Remove fluent interfaces from Currency

Arminek committed:
[Product] Removed fluent interfaces

NoResponseMate committed:
Remove fluent interfaces of Shippment component

Zales0123 committed:
[Review] Added listeners on Product and Customer deletion

Zales0123 committed:
[Review] PR review changes vol.1

HerbMat committed:
Remove fluent interfaces in Taxation

Commit:
Removed fluent interfaces

HerbMat committed:
Delete commented line in CoreContext

Commit:
[Promotion] Remove all fluent interfaces

HerbMat committed:
Delete commented code in HookContext

Commit:
Add check to customer context

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets |
| License       | MIT
| Doc PR        |

In one part of our site we don't need sylius autentication, just basic http with a generic username and password (which by default uses a Symfony user), so the Customer context fails when calling `->getCustomer()`.
Adding this check allows us to use basic http autentication.

NoResponseMate committed:
Removed fluent interfaces from Attribute class

Commit:
Remove unused dependencies

Zales0123 committed:
[Review] Added dynamic tags on ReviewDeleteListener

Also some changes/fixes in templates and crucial changes in review creation.

Commit:
Merge pull request #3334 from NeverResponse/remove-fluent-shipping

[Shipping] Removed fluent interfaces

Commit:
Merge pull request #3331 from gonzalovilaseca/patch-5

Remove unused dependencies

Commit:
Merge pull request #3330 from gonzalovilaseca/patch-4

Add check to customer context

Commit:
Merge pull request #3295 from NeverResponse/remove-fluent-attribute

[Attribute] Removed fluent interfaces

Commit:
Merge pull request #3252 from michalmarcinkowski/user-tweaks

[User] Update requirements and minor fixes

michalmarcinkowski committed:
Merge pull request #3291 from HerbMat/remove-fluent-interfaces-in-currency

[Currency] Remove fluent interfaces

michalmarcinkowski committed:
Merge pull request #3294 from 1XWP/SYLIUS-77-remove-fluent-interfaces-in-mailer-component

[Mailer] Removed fluent interfaces

michalmarcinkowski committed:
Merge pull request #3315 from TheMadeleine/translation-fluent-interfaces

[Translation] Remove fluent interfaces

michalmarcinkowski committed:
Merge pull request #3313 from HerbMat/remove-all-fluent-interfaces-in-taxation

[Taxation] Remove fluent interfaces

michalmarcinkowski committed:
Merge pull request #3327 from tuka217/promotion-fluent-interfaces-remove

[Promotion] Remove all fluent interfaces

michalmarcinkowski committed:
Merge pull request #3328 from Arminek/SYLIUS-82-product-component

[Product] Removed fluent interfaces

michalmarcinkowski committed:
Merge pull request #3322 from HerbMat/delete-unused-line-in-core-context

[Core] Delete commented line in CoreContext

michalmarcinkowski committed:
Merge pull request #3325 from HerbMat/delete-unused-code-in-hook-context

[Core] Delete commented code in HookContext
